front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy

```diff
@@ -44,6 +44,18 @@ class PipelineController {
     pipelineDAO.getPipelinesByApplication(application)
   }
 
+  @RequestMapping(value = 'getById/{id:.+}', method = RequestMethod.GET)
```

Comment:
I dont want a v2 or client impact in this CL. I'd like to converge S3 and
GCS first, and I'm leaving for vacation next week.
The proposal here is enabling future changes while still supporting current
usage patterns.

I am interpreting your previous reply (and out of band discussion earlier
this morning) as the following:

Current API

global scope means multiple instances across multiple apps
application scope means multiple instances in app, instance = one instance)
instance scope means one specific instance


global list       GET /pipelines
global save       POST /pipelines
global update     POST /pipelines/batchUpdate
instance delete   DELETE /pipelines/deleteById/{id}
instance delete   DELETE /pipelines/{application}/{pipeline}
instance rename   POST /pipelines/move

instance history  GET  /pipelines/{id}/history
application list  GET /pipelines/{application}


Proposed API (in this PR)

DEPRECATE (keep around but discourage use)
  instance delete   DELETE /pipelines/deleteById/{id}
 instance history  GET  /pipelines/{id}/history
 application list  GET /pipelines/{application}

ADD  (in this PR)
   instance update   PUT /pipelines/{id}
 instance get      GET /pipelines?id={id}
 application list  GET /pipelines?application={application}

And in addition Consider Extending the API  (in future PR)

  instance version-meta   GET /pipelines/versions/{id}
  instance version        GET /pipelines?id={id}&version={version}
  instance restore        PUT /pipelines/restore/{id}/{version}
                                                 or    PUT
/pipelines/{id}?version={version}
                                                 but this doesnt take a
payload like update does so might be dangerous.
    application  some way to list deleted pipeline ids



It would be nice if Applications (and notifications) were consistent with
these. Notifications look a bit fishy so am putting that off for now (but
raising them since I'm asking you to confirm how this can end up
consistent, which will need to accomodate them).
Applications have .../applications/name/{application} for methods that
operate on instances where .../applications and
.../applications/batchUpdate operate in the equivalent of global. If that
is the case, I propose pipelines use /id/ for the instance APIs.

ADD  (in this PR)
  instance update   PUT /pipelines/id/{id}
 instance get      GET /pipelines/id/{id}
 application list  GET /pipelines/application/{application}

This conflicts with  the existing GET /pipelines/{id}/history. Until it can
be removed, the implementations of the ambiguous APIs can check their
parameters for "id" or "history" depending on which, and call the right
handling. Unless actually specifying this causes a error from the ambiguous
specification.

The proposed extension would also use the /id/ paths, except for restore.



Thoughts?





On Wed, Jun 22, 2016 at 1:59 PM, Adam Jordens <notifications@github.com>
wrote:

> In
> front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy
> <https://github.com/spinnaker/front50/pull/98#discussion_r68102938>:
>
> > @@ -44,6 +44,18 @@ class PipelineController {
> >      pipelineDAO.getPipelinesByApplication(application)
> >    }
> >
> > +  @RequestMapping(value = 'getById/{id:.+}', method = RequestMethod.GET)
>
> I believe the API should be similar to:
>
> GET /pipelines/{id}
> GET /pipelines/{id}/history
> DELETE /pipelines/{id}
> PUT /pipelines/{id}
>
> POST /pipelines
> GET /pipelines
> GET /pipelines?name=abc
> GET /pipelines?application=xyz
>
> In the previous comment I included {application} because /pipeline/{id} is
> overloaded to the extent of not being usable (*if it's an application, we
> expect a collection ... if it's an id, we expect a single result*).
>
> We could bite bullet and version since what is being talked about here is
> a somewhat significant change (/v2/pipelines?) affecting multiple clients.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/front50/pull/98/files/15c62c9b564dc818eecfd59c4d45ef4a0889b2bf#r68102938>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEKkGBmTCbt-vn3EjPgM3_8DA5pqZeAtks5qOXf8gaJpZM4I58yx>
> .
>